### PR TITLE
Fix layer.clone bug

### PR DESCRIPTION
**Pull request by Pessimistress**
#### Background
`layer.clone()` always results in `data` prop being an empty array.

#### Change List
- Fix async prop cloning
- Add test case

### CODE REVIEW

File: modules/core/src/lifecycle/component.js [hunk 1]
```diff
@@ -32,7 +32,20 @@ export default class Component {
...
+
+    for (const key in props._asyncPropDefaultValues) {
+      if (key in props._asyncPropResolvedValues) {
+        asyncProps[key] = props._asyncPropResolvedValues[key];
+      } else if (key in props._asyncPropOriginalValues) {
+        asyncProps[key] = props._asyncPropOriginalValues[key];
+      }
+    }
+
+    return new this.constructor(Object.assign({}, props, asyncProps, newProps));
```
Comment (ibgreen) [inline modules/core/src/lifecycle/component.js]: Nit: Not part of the scope of the change in this PR, feel free to ignore:

When I read this, I suspect that `clone` is not participating in the optimization that avoids/minimizes the repeated copying of all props using `Object.assign`?

Would ` return new this.constructor(props, asyncProps, newProps);` work instead? I think that would avoid one round of `Object.assign`?

Maybe add a bench case for clone together with the existing layer instantiation cases and then do that change? At least so we know if `clone` has a similar perf as new.

File: modules/core/src/lifecycle/component.js [hunk 2]
```diff
@@ -32,7 +32,20 @@ export default class Component {
 
   // clone this layer with modified props
   clone(newProps) {
-    return new this.constructor(Object.assign({}, this.props, newProps));
+    const {props} = this;
+
+    // Async props cannot be copied with Object.assign. Extract the plain value.
```
Comment (ibgreen) [inline modules/core/src/lifecycle/component.js]: The last part of the comment doesn't seem helpful...

Extract the plain value. => Copy them separately?

File: modules/core/src/lifecycle/component.js [hunk 3]
```diff
@@ -32,7 +32,20 @@ export default class Component {
...
+    const asyncProps = {};
+
+    for (const key in props._asyncPropDefaultValues) {
```
Comment (gnavvy) [inline modules/core/src/lifecycle/component.js]: nit: // refer the definitions of `_asyncProp*Values` in `create-props.js`.

File: modules/core/src/lifecycle/component.js (hunk 1 above)
Comment (Pessimistress) [reply modules/core/src/lifecycle/component.js]: Unfortunately, a bunch of core layers ([example](https://github.com/uber/deck.gl/blob/master/modules/layers/src/line-layer/line-layer.js#L41)) do not support constructor with multiple arguments, which is another bug I'll fix. Potentially any custom layers created by the user may override `constructor` incorrectly too.

While this solution might be slower, it is guaranteed to clone a layer safely.

**Pull request merged**